### PR TITLE
Bundle size tracking

### PR DESCRIPTION
**Pull request by ericclemmons**
_Issue #, if available:_  https://www.pivotaltracker.com/story/show/174995888

This PR introduces the foundation to check real-world application bundle sizes within PRs. 

See https://ja2r7.app.goo.gl/6PJJdRZMA6mz3ZWH7 for an example of the results & the GitHub status checks below 👇 

Future improvements would be:

- [ ] Set thresholds on the net ± impact
- [ ] Update the status check with results (https://github.com/mitchheddles/github-action-status-check)

## Naive Implementation

We _could_ simply report on the fully-bundled cost of our own code (i.e. the size of a package if all source was concatenated, excluding dependencies):

```
PASS packages/amazon-cognito-identity-js/dist/amazon-cognito-identity.min.js: 24.3KB < 400KB (gzip)
PASS packages/amplify-ui-react/dist/aws-amplify-react.min.js: 55.31KB < 400KB (gzip)
PASS packages/amplify-ui/dist/aws-amplify-ui.min.js: 1.53KB < 400KB (gzip)
PASS packages/analytics/dist/aws-amplify-analytics.min.js: 55.34KB < 400KB (gzip)
PASS packages/api-graphql/dist/aws-amplify-api-graphql.min.js: 37.89KB < 400KB (gzip)
PASS packages/api-rest/dist/aws-amplify-api-rest.min.js: 12.57KB < 400KB (gzip)
PASS packages/api/dist/aws-amplify-api.min.js: 38.91KB < 400KB (gzip)
PASS packages/auth/dist/aws-amplify-auth.min.js: 164.98KB < 400KB (gzip)
PASS packages/aws-amplify-react/dist/aws-amplify-react.min.js: 50.56KB < 400KB (gzip)
PASS packages/aws-amplify-vue/dist/aws-amplify-vue.umd.min.js: 48.02KB < 400KB (gzip)
PASS packages/aws-amplify/dist/aws-amplify.min.js: 346.76KB < 400KB (gzip)
PASS packages/cache/dist/aws-amplify-cache.min.js: 3.97KB < 400KB (gzip)
PASS packages/core/dist/aws-amplify-core.min.js: 42.63KB < 400KB (gzip)
PASS packages/datastore/dist/aws-amplify-datastore.min.js: 255.19KB < 400KB (gzip)
PASS packages/interactions/dist/aws-amplify-interactions.min.js: 23.86KB < 400KB (gzip)
PASS packages/predictions/dist/aws-amplify-predictions.min.js: 101.6KB < 400KB (gzip)
PASS packages/pubsub/dist/aws-amplify-pubsub.min.js: 29.47KB < 400KB (gzip)
PASS packages/storage/dist/aws-amplify-storage.min.js: 61.42KB < 400KB (gzip)
PASS packages/xr/dist/aws-amplify-xr.min.js: 3.87KB < 400KB (gzip)
```

However, **this doesn't match real-world experience with Amplify**.

## Real-world Bundles

Instead, this PR introduces multiple tests around specific usage of Amplify to test how bundles (e.g. `webpack`) handle tree-shaking:

```js
import { Amplify, Auth } from "aws-amplify"

Amplify.configure()
Auth.signIn()
```

In this example, `Amplify` and `Auth` packages will be included in the bundle output, as well as any 3rd party dependencies they depend on.

**With this approach, we can ensure that any mistakes with tree-shaking or erroneous-included dependencies are visible in PRs**.

For example, `webpack4` includes a `crypto` polyfill:

```
PASS dist/amazon-cognito-identity-js.js.min.js: 149.83KB < 150KB (gzip)
PASS dist/Amplify.js.min.js: 1.55KB < 2KB (gzip)
PASS dist/Amplify+Auth.js.min.js: 195.22KB < 200KB (gzip)
```

By applying these changes, customers can disable it:

```diff
--- a/.github/actions/bundle-size-action/webpack4/webpack.config.js
+++ b/.github/actions/bundle-size-action/webpack4/webpack.config.js
@@ -15,4 +15,8 @@ module.exports = {
        },

        target: 'web',
+
+       externals: {
+               crypto: 'crypto',
+       },
 };
```

**Resulting in `125kB` in savings**:

```
PASS dist/amazon-cognito-identity-js.js.min.js: 23.44KB < 150KB (gzip)
PASS dist/Amplify.js.min.js: 1.55KB < 2KB (gzip)
PASS dist/Amplify+Auth.js.min.js: 68.7KB < 200KB (gzip)
```

In these situations, we can start by setting thresholds with this PR, then eventually track the net difference in bundle sizes per-PR.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ericclemmons** [comment]
Hmm, this comes _pretty close_, but I was hoping to get tighter integration with the check response:

```
yarn run v1.22.5
$ /home/runner/work/amplify-js/amplify-js/node_modules/.bin/bundlewatch
Warning:  Some CI configuration options were not found (githubAccessToken):
    bundlewatch will be unable to report build status, or save comparison data
    Learn more at: https://bundlewatch.io/
        

Result breakdown at: https://ja2r7.app.goo.gl/X9Q8WZtaNgh99gKE7

PASS packages/core/dist/aws-amplify-core.min.js: 42.63KB < 400KB (gzip)

bundlewatch PASS
Everything is in check (+42.63KB, -0B)
```

So we can tune this to pass/fail and report (e.g. https://ja2r7.app.goo.gl/AC8YjgbXhkggFZoK7), but I would like to improve _visibility_ into bundle size changes.

Maybe this could be enough for a proof-of-concept, and I can clamp each bundle output to the nearest ~25Kb.  This would at least catch major increases in bundle-sizes, but wouldn't really catch _real-world_ increases that I'd prefer.

A future PR could go further and have more integration-like tests:

```js
// test/integration/bundlewatch/AmplifyAuth.ts
import { Amplify, Auth } from "aws-amplify"

export { Amplify, Auth }
```

Then, do a bundle-size comparison of the result of `rollup` / `webpack` / `microbundle` / `react-scripts` to validate _real-world_ usage of Amplify doesn't get bloated via 3rd party dependencies or broken tree-shaking.

File: package.json [hunk 1]
```diff
@@ -98,6 +99,14 @@
 		"terser": "4.6.7",
 		"npm-packlist": "1.1.12"
 	},
+	"bundlewatch": {
+		"files": [
+			{
+				"path": "packages/*/dist/*.min.js",
+				"maxSize": "400kB"
```
Comment (ericclemmons) [inline package.json:106]: Before merging, we would want to tune each package to a `maxSize` based on https://ja2r7.app.goo.gl/6PJJdRZMA6mz3ZWH7.

Comment (sammartinez) [reply package.json:106]: Does it work from an overall or each individual is the way to go?

File: package.json [hunk 2]
```diff
@@ -98,6 +99,14 @@
 		"terser": "4.6.7",
 		"npm-packlist": "1.1.12"
 	},
+	"bundlewatch": {
```
Comment (sammartinez) [inline package.json:102]: License is MIT, I believe we need to double check if we need to do anything internally

**Comment by sammartinez** [comment]
As we discussed offline, once we have the limits in place on the packages, we can look to get this in

File: .github/workflows/bundle-size.yml [hunk 3]
```diff
@@ -31,6 +31,8 @@ jobs:
       # TODO Publish to verdaccio
       # https://verdaccio.org/docs/en/github-actions
       # But does this persist for the next step?
+      - name: Publish with Verdaccio
+        uses: verdaccio/github-actions/publish@master
```
Comment (sammartinez) [inline .github/workflows/bundle-size.yml:35]: Does this need to be `@main`?

**Comment by ericclemmons** [comment]
I'm trying to understand why Next.js' stats check works with PRs, while our doesn't:

- https://github.com/vercel/next.js/pull/21123#issuecomment-760289510
- https://github.com/vercel/next.js/blob/dc87a8436bd978a8cf9e605da171897400b19743/.github/actions/next-stats-action/src/prepare/action-info.js#L17

And I'm passing it via:
- https://github.com/ericclemmons/amplify-js/blob/174995888-bundle-size/.github/workflows/bundle-size.yml#L54

There's an open issue indicated that secrets aren't available to forks, so how does theirs work 🤔 
> https://github.community/t/make-secrets-available-to-builds-of-forks/16166/33

**Review by iartemiev — APPROVED** [review]
Faux approval (testing PR functionality). Still need to re-review

**Comment by ericclemmons** [comment]
@sammartinez This check will only happen when there's an approval.

This is what it'll look like:
> Screen Shot 2021-01-14 at 4 55 21 PM

**Comment by ericclemmons** [comment]
I added another test based on https://github.com/aws-amplify/amplify-js/issues/7570

**Review by amhinson — APPROVED** [review]
Everything looks good to me 👍  Approving to see the GitHub Actions in use.

File: .github/actions/bundle-size-action/next/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+This is a [Next.js](https://nextjs.org/) project bootstrapped with [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app).
```
Comment (sammartinez) [inline .github/actions/bundle-size-action/next/README.md:1]: Do we need this?

File: .github/actions/bundle-size-action/next/package.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    "prebuild": "rimraf .next",
+    "build": "next build",
+    "start": "next start",
+    "precalculate": "$_ run build",
+    "calculate": "bundlewatch"
+  },
+  "bundlewatch": {
+    "files": [
+      {
+        "path": ".next/static/chunks/pages/index-d6a61784a8ae4ca7feaf.js",
```
Comment (sammartinez) [inline .github/actions/bundle-size-action/next/package.json:16]: Will the chunk change?

File: .github/workflows/bundle-size.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    branches: [main]
+
+jobs:
+  bundle-size:
+    # if: github.event.review.state == 'approved'
+    name: Calculate
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        stack: [next, webpack4, webpack]
```
Comment (sammartinez) [inline .github/workflows/bundle-size.yml:20]: Future thought is also seeing about Rollup

**Review by sammartinez — APPROVED** [review]
LGTM 🌮  Overall. Couple call outs but they are minor.

**Review by sammartinez — APPROVED** [review]
LGTM 🌮

**Pull request merged**